File: apps/server/src/source.ts
import type {
  ChangedFile,
  DiffMode,
  FilePatch,
  RepositoryDiff,
} from "@servediff/shared";

export interface DiffSource {
  root: string;
  kind: "local" | "stdin";
  scopes: readonly DiffMode[];
  live: boolean;
  snapshot(mode: DiffMode): Promise<RepositoryDiff>;
  patch(
    mode: DiffMode,
    file: ChangedFile,
    head: string | null,
  ): Promise<FilePatch>;
  contents(
    mode: DiffMode,
    file: ChangedFile,
    head: string | null,
  ): Promise<{ before: string; after: string }>;
}

export class RequestError extends Error {
  status: number;
  constructor(status: number, message: string) {
    super(message);
    this.status = status;
  }
}

File: internal/diffsource/source.go
package diffsource

import (
	"context"
	"fmt"

	"github.com/flexdinesh/servediff/internal/review"
)

type Source interface {
	Root() string
	Kind() string
	Support() Support
	Snapshot(context.Context, review.DiffMode) (review.RepositoryDiff, error)
	Patch(context.Context, review.DiffMode, review.ChangedFile, *string) (review.FilePatch, error)
	Contents(context.Context, review.DiffMode, review.ChangedFile, *string) (review.FileContents, error)
}

type Support struct {
	Scopes          []review.DiffMode
	Refresh         bool
	StagingMetadata bool
	FileContents    bool
}

type RequestError struct {
	Status int
	Detail string
}

func (err *RequestError) Error() string { return err.Detail }

func Error(status int, format string, values ...any) error {
	return &RequestError{Status: status, Detail: fmt.Sprintf(format, values...)}
}

func Message(value string) *string { return &value }
